src/main/java/net/dv8tion/jda/api/audio/AudioSendHandler.java

```diff
@@ -50,9 +52,14 @@
      * <br>Defined by: {@link net.dv8tion.jda.api.audio.AudioSendHandler#INPUT_FORMAT AudioSendHandler.INPUT_FORMAT}.
      * <br>If {@link #isOpus()} is set to return true, then it should be in pre-encoded Opus format instead.
      *
-     * @return Should return a byte[] containing 20 Milliseconds of audio.
+     * @return Should return a {@link java.nio.ByteBuffer} containing 20 Milliseconds of audio.
+     *
+     * @see    #isOpus()
+     * @see    #canProvide()
+     * @see    java.nio.ByteBuffer#allocate(int)
+     * @see    java.nio.ByteBuffer#wrap(byte[])
      */
-    byte[] provide20MsAudio();
+    ByteBuffer provide20MsAudio();
```

Comment:
Shouldn't this method have a ByteBuffer as an argument that the audio data should be written to?
Returning a buffer looks off because how does the implementor know when the buffer it returns here is safe to be used again? Or am I on the wrong track about how this would be used in an implementation with the goal of minimal allocations?

Reply:
If you would do provide(ByteBuffer) you get additional overhead for copying over data. It would also severely limit usability since the buffer could not be re-allocated correctly. The buffer should be updated by the next provide cycle. Meaning canProvide signals the buffer was sent and the next one will be requested. If you are worried about concurrency you an always use `buffer.duplicate()` instead or have a queue of buffers. Since we immediately read the buffer and write the encrypted data into a different buffer of the packet provider it should almost immediately be usable again.

Reply:
Additionally, the goal of these changes was *not* to force users into single-allocation which your suggestion would imply. The goal was to make it possible in the first place. Since returning a buffer with a position and limit you can create a fitting requirement without re-allocation compared to using a byte[]. This is due to the fact that byte[] needs to be allocated with the exact size needed for the audio frame while the buffer can be adjusted for it without any allocations. To achieve 0 allocations you simply need to have one buffer that is filled by canProvide() and then returned by provide() as was shown by the examples in the comments of this PR.

Reply:
Thanks for clarifying!